### ISSUE TITLE
Factor columns sometimes not appearing in selector

### ISSUE THREAD

**Issue by dannyparsons** [africanmathsinitiative/R-Instat#5302]
**Describe the bug**
I haven't been able to reproduce this but noticed several times that the selector sometimes doesn't display the factor columns. I've only noticed this after running several dialogs first. Strangely if you convert a factor column to factor it then reappears in the selector. I'm not sure if it effects all dialogs or just some or its related to certain kinds of data files.

**To Reproduce**
@rdstern has also noticed this and may be able to be more specific.

If someone can first get a reproduceable example that would help greatly in resolving this.

Ideally someone can solve it too but getting a reproduceable example first will really help in getting to the cause of this.
@africanmathsinitiative/developers this is quite a serious problem so it should be a priority to resolve.

**Comment by rdstern** [comment]
I think the problem is with the **Climatic > Dates > Infill Missing Dates** dialogue.
I opened the Instat data climatic, ghana.
a) If I go directly to the Define climatic data the station is fine.  It always appears.
b) If I then go to the Infill missing - there are some to be infilled - and then return it seems ok.  But if I press reset it then can't find the station.

I wonder if the infill has changed the type of the station to character and this has not changed in the grid.  If so, it would explain why the station is suddently not recognised.

**Comment by Wycklife** [comment]
@rdstern I don't seems to be getting the same problem. If I go to the infill dialog, only Station appears in the selector as a factor in the By factor receiver. When I reset the dialog, the station disappears from the factor column receiver but present in the selector. When I go back to the original dialog(Define Climatic data). the station is still present. Please could we discuss this?

**Comment by Wycklife** [comment]
Another observation is that, after resetting the dialog, the station disappears also from the metadata.

**Comment by rdstern** [comment]
Oh good.  This should make the discovery of the bug easier, noting that the problem is actually within the Infill dialogue.  Can you see where the R code in that dialogue is changing the station?  I assume it is then changed into a character variable and is therefore no longer a factor.  So possibly simply adding an as.factor to that expression may be needed?

**Comment by shadrackkibet** [comment]
This Bug is a serious one and also disturbing. It is worth someone spending time to look into it.This might be related to issue #4127 . I will also try to replicate it next week if i get some free time.

**Comment by Wycklife** [comment]
I have added as.factor to the code on the script window but it produces an error.
data_book$infill_missing_dates(data_name="ghana", date_name="date", factors=as.factor ("station"))
![image](https://user-images.githubusercontent.com/33898404/56360441-530cc500-61ed-11e9-9925-37c35c769fe4.png)

When I convert the factor column to factor after resetting the dialog, the station seems to appear on the Selector. But when I convert the factor column to factor before opening the dialog, the problem still persist when I reset the dialog.

**Comment by dannyparsons** [comment]
I suspect this might be an issue with the metadata not recognising it as a factor. This would need to be look at within the `data_book` code.

Also, you can't do `as.factor("station")` because its not the column, it's the string `"station"` which is passed in to the `data_book`.

**Comment by Wycklife** [comment]
@dannyparsons Could you please help me fix this issue?

**Comment by dannyparsons** [comment]
Yes I need to look into this. I think we discovered that any time rows are added (e.g. infilling) or deleted (delete a row) then this problem happens. So I think something we will find in the metadata.
